File: src/gitCrypt/gitCryptCli.ts
import { spawn } from 'node:child_process';
import { chmod, readdir } from 'node:fs/promises';
import * as path from 'node:path';
import { GitCryptRepositoryStatus } from './types';

const MAX_COMMAND_OUTPUT_BYTES = 1024 * 1024;

export class GitCryptCommandError extends Error {
  public constructor(
    message: string,
    public readonly unavailable: boolean,
    public readonly exitCode?: number,
  ) {
    super(message);
    this.name = 'GitCryptCommandError';
  }
}

export interface CommandResult {
  readonly stdout: string;
  readonly stderr: string;
}

export interface CommandExecutor {
  run(executable: string, args: readonly string[], cwd: string): Promise<CommandResult>;
}

export class ProcessCommandExecutor implements CommandExecutor {
  public run(executable: string, args: readonly string[], cwd: string): Promise<CommandResult> {
    return new Promise((resolve, reject) => {
      const child = spawn(executable, args, {
        cwd,
        env: { ...process.env, LC_ALL: 'C' },
        stdio: ['ignore', 'pipe', 'pipe'],
        windowsHide: true,
      });
      const stdout: Buffer[] = [];
      const stderr: Buffer[] = [];
      let outputBytes = 0;
      let settled = false;

      const fail = (error: GitCryptCommandError): void => {
        if (settled) {
          return;
        }
        settled = true;
        child.kill();
        reject(error);
      };

      child.once('error', (error: NodeJS.ErrnoException) => {
        fail(
          new GitCryptCommandError(
            error.code === 'ENOENT'
              ? 'git-crypt is not installed or is not available on PATH.'
              : error.message,
            error.code === 'ENOENT',
          ),
        );
      });

      const collect = (target: Buffer[]) => (chunk: Buffer): void => {
        outputBytes += chunk.length;
        if (outputBytes > MAX_COMMAND_OUTPUT_BYTES) {
          fail(new GitCryptCommandError('git-crypt output exceeded the safety limit.', false));
          return;
        }
        target.push(chunk);
      };
      child.stdout.on('data', collect(stdout));
      child.stderr.on('data', collect(stderr));

      child.once('close', (code) => {
        if (settled) {
          return;
        }
        settled = true;
        const result = {
          stdout: Buffer.concat(stdout).toString('utf8'),
          stderr: Buffer.concat(stderr).toString('utf8'),
        };
        if (code !== 0) {
          reject(
            new GitCryptCommandError(
              result.stderr.trim() || `git-crypt exited with code ${code ?? 'unknown'}.`,
              false,
              code ?? undefined,
            ),
          );
          return;
        }
        resolve(result);
      });
    });
  }
}

/** Performs explicit git-crypt operations without reading key or protected file contents. */
export class GitCryptCli {
  public constructor(
    private readonly executable = 'git-crypt',
    private readonly executor: CommandExecutor = new ProcessCommandExecutor(),
  ) {}

  public async inspect(
    repositoryRoot: string,
    gitDir: string,
    hasProtectedFiles: boolean,
  ): Promise<GitCryptRepositoryStatus> {
    const installedKeyCount = await countInstalledKeys(gitDir);
    try {
      const result = await this.executor.run(this.executable, ['--version'], repositoryRoot);
      return {
        available: true,
        version: result.stdout.trim() || result.stderr.trim() || undefined,
        localState:
          installedKeyCount > 0
            ? 'unlocked'
            : hasProtectedFiles
              ? 'locked'
              : 'not-initialized',
        installedKeyCount,
      };
    } catch (error) {
      if (error instanceof GitCryptCommandError && error.unavailable) {
        return {
          available: false,
          localState:
            installedKeyCount > 0
              ? 'unlocked'
              : hasProtectedFiles
                ? 'locked'
                : 'not-initialized',
          installedKeyCount,
        };
      }
      throw error;
    }
  }

  public async unlockWithKey(repositoryRoot: string, keyFile: string): Promise<void> {
    const absoluteKeyFile = path.resolve(keyFile);
    try {
      await this.executor.run(this.executable, ['unlock', absoluteKeyFile], repositoryRoot);
    } catch (error) {
      throw maskPathInError(error, absoluteKeyFile, '<key file>');
    }
  }

  public async initializeRepository(repositoryRoot: string): Promise<void> {
    await this.executor.run(this.executable, ['init'], repositoryRoot);
  }

  public async unlockWithGpg(repositoryRoot: string): Promise<void> {
    await this.executor.run(this.executable, ['unlock'], repositoryRoot);
  }

  public async lockRepository(repositoryRoot: string): Promise<void> {
    await this.executor.run(this.executable, ['lock', '--all'], repositoryRoot);
  }

  public async addGpgUser(repositoryRoot: string, gpgUserId: string): Promise<void> {
    const normalizedUserId = gpgUserId.trim();
    if (normalizedUserId.length === 0) {
      throw new GitCryptCommandError('A GPG user ID is required.', false);
    }
    if (normalizedUserId.startsWith('-') || /[\0\r\n]/u.test(normalizedUserId)) {
      throw new GitCryptCommandError('The GPG user ID contains unsupported characters.', false);
    }
    try {
      await this.executor.run(
        this.executable,
        ['add-gpg-user', '--no-commit', normalizedUserId],
        repositoryRoot,
      );
    } catch (error) {
      throw maskPathInError(error, normalizedUserId, '<GPG user ID>');
    }
  }

  public async exportKey(repositoryRoot: string, destination: string): Promise<boolean> {
    const absoluteDestination = path.resolve(destination);
    try {
      await this.executor.run(this.executable, ['export-key', absoluteDestination], repositoryRoot);
      if (process.platform !== 'win32') {
        try {
          await chmod(absoluteDestination, 0o600);
          return true;
        } catch {
          return false;
        }
      }
      return false;
    } catch (error) {
      throw maskPathInError(error, absoluteDestination, '<export destination>');
    }
  }
}

async function countInstalledKeys(gitDir: string): Promise<number> {
  try {
    const entries = await readdir(path.join(gitDir, 'git-crypt', 'keys'), {
      withFileTypes: true,
    });
    return entries.filter((entry) => entry.isFile() || entry.isDirectory()).length;
  } catch (error) {
    const code = (error as NodeJS.ErrnoException).code;
    if (code === 'ENOENT' || code === 'ENOTDIR') {
      return 0;
    }
    throw error;
  }
}

function maskPathInError(error: unknown, sensitivePath: string, replacement: string): Error {
  if (!(error instanceof Error)) {
    return new GitCryptCommandError('Unknown git-crypt error.', false);
  }
  const escapedPath = sensitivePath.replace(/[.*+?^${}()|[\]\\]/gu, '\\$&');
  const message = error.message.replace(new RegExp(escapedPath, 'gu'), replacement);
  if (error instanceof GitCryptCommandError) {
    return new GitCryptCommandError(message, error.unavailable, error.exitCode);
  }
  return new Error(message);
}

File: intellij-plugin/src/main/kotlin/com/ymkang/gitcryptexplorer/core/GitCryptCli.kt
package com.ymkang.gitcryptexplorer.core

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.attribute.PosixFilePermissions
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

private const val MAX_COMMAND_OUTPUT_BYTES = 1024L * 1024
private val CLI_IO_EXECUTOR = Executors.newCachedThreadPool()

class GitCryptCli(private val executable: String = resolveGitCryptExecutable()) {
    fun inspect(repositoryRoot: Path, gitDir: Path, hasProtectedFiles: Boolean): GitCryptRepositoryStatus {
        val installedKeyCount = countInstalledKeys(gitDir)
        return try {
            val result = run(listOf("--version"), repositoryRoot)
            GitCryptRepositoryStatus(true, result.stdout.trim().ifEmpty { result.stderr.trim() }.ifEmpty { null }, localState(installedKeyCount, hasProtectedFiles), installedKeyCount)
        } catch (error: GitCryptCommandException) {
            if (!error.unavailable) throw error
            GitCryptRepositoryStatus(false, localState = localState(installedKeyCount, hasProtectedFiles), installedKeyCount = installedKeyCount)
        }
    }

    fun unlockWithKey(repositoryRoot: Path, keyFile: Path) = runMasked(listOf("unlock", keyFile.toAbsolutePath().toString()), repositoryRoot, keyFile, "<key file>")
    fun initializeRepository(repositoryRoot: Path) { run(listOf("init"), repositoryRoot) }
    fun unlockWithGpg(repositoryRoot: Path) { run(listOf("unlock"), repositoryRoot) }
    fun lockRepository(repositoryRoot: Path) { run(listOf("lock", "--all"), repositoryRoot) }

    fun addGpgUser(repositoryRoot: Path, userId: String) {
        val normalized = userId.trim()
        if (normalized.isEmpty()) throw GitCryptCommandException("A GPG user ID is required.")
        if (normalized.startsWith('-') || normalized.any { it == '\u0000' || it == '\r' || it == '\n' }) {
            throw GitCryptCommandException("The GPG user ID contains unsupported characters.")
        }
        runMasked(listOf("add-gpg-user", "--no-commit", normalized), repositoryRoot, Path.of(normalized), "<GPG user ID>")
    }

    fun exportKey(repositoryRoot: Path, destination: Path): Boolean {
        val absolute = destination.toAbsolutePath().normalize()
        runMasked(listOf("export-key", absolute.toString()), repositoryRoot, absolute, "<export destination>")
        if (System.getProperty("os.name").contains("win", ignoreCase = true)) return false
        return try {
            Files.setPosixFilePermissions(absolute, PosixFilePermissions.fromString("rw-------"))
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun runMasked(args: List<String>, cwd: Path, sensitive: Path, replacement: String) {
        try {
            run(args, cwd)
        } catch (error: GitCryptCommandException) {
            throw GitCryptCommandException(error.message.orEmpty().replace(sensitive.toString(), replacement), error.unavailable, error.exitCode)
        }
    }

    private fun run(args: List<String>, cwd: Path): CommandResult {
        val process = try {
            ProcessBuilder(executable, *args.toTypedArray()).directory(cwd.toFile()).apply { environment()["LC_ALL"] = "C" }.start()
        } catch (error: java.io.IOException) {
            throw GitCryptCommandException("git-crypt is not installed or is not available on PATH.", error.message?.contains("Cannot run program") == true)
        }
        val stdoutFuture = CompletableFuture.supplyAsync({ process.inputStream.readLimited(MAX_COMMAND_OUTPUT_BYTES) }, CLI_IO_EXECUTOR)
        val stderrFuture = CompletableFuture.supplyAsync({ process.errorStream.readLimited(MAX_COMMAND_OUTPUT_BYTES) }, CLI_IO_EXECUTOR)
        process.outputStream.close()
        if (!process.waitFor(30, TimeUnit.SECONDS)) {
            process.destroyForcibly()
            throw GitCryptCommandException("git-crypt command timed out.")
        }
        val stdout = joinOutput(stdoutFuture).toString(Charsets.UTF_8)
        val stderr = joinOutput(stderrFuture).toString(Charsets.UTF_8)
        val exit = process.exitValue()
        if (exit != 0) throw GitCryptCommandException(stderr.trim().ifEmpty { "git-crypt exited with code $exit." }, exitCode = exit)
        return CommandResult(stdout, stderr)
    }

    private fun joinOutput(future: CompletableFuture<ByteArray>): ByteArray = try {
        future.join()
    } catch (error: CompletionException) {
        throw (error.cause as? GitCryptCommandException ?: GitCryptCommandException(error.message.orEmpty()))
    }

    private fun countInstalledKeys(gitDir: Path): Int = try {
        Files.list(gitDir.resolve("git-crypt").resolve("keys")).use { stream -> stream.count().toInt() }
    } catch (error: java.io.IOException) {
        if (error is java.nio.file.NoSuchFileException || error is java.nio.file.NotDirectoryException) 0 else throw error
    }

    private fun localState(keyCount: Int, hasProtectedFiles: Boolean): GitCryptLocalState = when {
        keyCount > 0 -> GitCryptLocalState.UNLOCKED
        hasProtectedFiles -> GitCryptLocalState.LOCKED
        else -> GitCryptLocalState.NOT_INITIALIZED
    }

    private data class CommandResult(val stdout: String, val stderr: String)
}

/**
 * GUI-launched JetBrains IDEs do not always inherit the user's shell PATH.
 * Prefer the normal PATH lookup, then check the standard Homebrew and system
 * locations so macOS installations work without launching the IDE from a shell.
 */
private fun resolveGitCryptExecutable(): String {
    val candidates = linkedSetOf<String>()
    System.getenv("PATH")
        ?.split(java.io.File.pathSeparator)
        ?.filter(String::isNotBlank)
        ?.forEach { directory ->
            candidates += Path.of(directory).resolve("git-crypt").toString()
            if (System.getProperty("os.name").contains("win", ignoreCase = true)) {
                candidates += Path.of(directory).resolve("git-crypt.exe").toString()
            }
        }
    candidates += listOf(
        "/opt/homebrew/bin/git-crypt",
        "/opt/homebrew/opt/git-crypt/bin/git-crypt",
        "/usr/local/bin/git-crypt",
        "/usr/local/opt/git-crypt/bin/git-crypt",
        "/usr/bin/git-crypt",
        "/bin/git-crypt",
    )
    return candidates.firstOrNull { candidate ->
        val path = Path.of(candidate)
        Files.isRegularFile(path) && Files.isExecutable(path)
    } ?: "git-crypt"
}

private fun java.io.InputStream.readLimited(limit: Long): ByteArray {
    val output = java.io.ByteArrayOutputStream()
    val buffer = ByteArray(8192)
    var total = 0L
    while (true) {
        val read = read(buffer)
        if (read < 0) break
        total += read
        if (total > limit) throw GitCryptCommandException("git-crypt output exceeded the safety limit.")
        output.write(buffer, 0, read)
    }
    return output.toByteArray()
}
